__tests__/integration.js

```diff
@@ -0,0 +1,391 @@
+      succeed
+    })
+  })
+
+  test('PUT JSON', (done) => {
+    const succeed = response => {
+      delete response.headers.date
+      expect(response).toEqual(makeResponse({
+        "body": '{"id":2,"name":"Samuel"}',
+        "headers": {
+          "content-length": "24",
+          "etag": "W/\"18-uGyzhJdtXqacOe9WRxtXSNjIk5Q\"",
+        },
+        statusCode: 200
+      }))
+      done()
+    }
+    lambdaFunction.handler(makeEvent({
+      path: '/users/2',
+      httpMethod: 'PUT',
+      body: '{"name": "Samuel"}'
+    }), {
+      succeed
+    })
+  })
+
+  test('base64 encoded request', (done) => {
+    const succeed = response => {
+      delete response.headers.date
+      expect(response).toEqual(makeResponse({
+        "body": '{"id":2,"name":"Samuel"}',
+        "headers": {
+          "content-length": "24",
+          "etag": "W/\"18-uGyzhJdtXqacOe9WRxtXSNjIk5Q\"",
+        },
+        statusCode: 200
+      }))
+      done()
+    }
+    lambdaFunction.handler(makeEvent({
+      path: '/users/2',
+      httpMethod: 'PUT',
+      body: btoa('{"name": "Samuel"}'),
+      isBase64Encoded: true
+    }), {
+      succeed
+    })
+  })
+
+  test('forwardConnectionErrorResponseToApiGateway', (done) => {
+    const succeed = response => {
+      delete response.headers.date
+      expect(response).toEqual({
+        "body": "",
+        "headers": {},
+        statusCode: 502
+      })
+      done()
+    }
+    lambdaFunction.handler(makeEvent({
+      path: '/',
+      httpMethod: 'GET',
+      body: "{\"name\": \"Sam502\"}",
+      headers: {
+        'Content-Length': '-1'
+      }
+    }), {
+      succeed
+    })
+  })
+  
+  const mockApp = function (req, res) {
+    res.end('')
+  }
+
+  test('forwardLibraryErrorResponseToApiGateway', (done) => {
+    const succeed = response => {
+      expect(response).toEqual({
+        statusCode: 500,
+        body: '',
+        headers: {}
+      })
+      done()
+    }
+    awsServerlessExpress.proxy(server, null, {
+      succeed
+    })
+  })
+
+  test('serverListenCallback', (done) => {
+    const serverListenCallback = jest.fn()
+    const serverWithCallback = awsServerlessExpress.createServer(mockApp, serverListenCallback)
+    const succeed = response => {
+      expect(response.statusCode).toBe(200)
+      expect(serverListenCallback).toHaveBeenCalled()
+      serverWithCallback.close()
+      done()
+    }
+    awsServerlessExpress.proxy(serverWithCallback, makeEvent({}), {
+      succeed
+    })
+  })
+
+  test('server.onError EADDRINUSE', (done) => {
+    const serverWithSameSocketPath = awsServerlessExpress.createServer(mockApp)
+    serverWithSameSocketPath._socketPathSuffix = server._socketPathSuffix
+    const succeed = response => {
+      expect(response.statusCode).toBe(200)
+      done()
+      serverWithSameSocketPath.close()
+    }
+    awsServerlessExpress.proxy(serverWithSameSocketPath, makeEvent({}), {
+      succeed
+    })
+  })
+
+  test.skip('set-cookie')
```

Comment:
Another edge case test we can leave for another day. API Gateway does not have support for multiple headers with same name. We work around this by modifying the casing of the set-cookie header when we detect there are multiple.